storagemarket/impl/clientstates/client_states.go

```diff
@@ -46,9 +48,19 @@ func EnsureClientFunds(ctx fsm.Context, environment ClientDealEnvironment, deal
 		return ctx.Trigger(storagemarket.ClientEventEnsureFundsFailed, xerrors.Errorf("acquiring chain head: %w", err))
 	}
 
-	mcid, err := node.EnsureFunds(ctx.Context(), deal.Proposal.Client, deal.Proposal.Client, deal.Proposal.ClientBalanceRequirement(), tok)
+	requiredFunds, err := environment.DealFunds().Reserve(deal.Proposal.ClientBalanceRequirement())
+	if err != nil {
+		return ctx.Trigger(storagemarket.ClientEventEnsureFundsFailed, xerrors.Errorf("tracking deal funds: %w", err))
+	}
+
+	mcid, err := node.EnsureFunds(ctx.Context(), deal.Proposal.Client, deal.Proposal.Client, requiredFunds, tok)
 
 	if err != nil {
+		_, err2 := environment.DealFunds().Release(deal.Proposal.ClientBalanceRequirement())
```

Comment:
This needs to happen in the global fail state handler (FailDeal) - any error that happens between when EnsureFunds is called and the deal is published need to release the funds

Reply:
to know whether to release in FailDeal may mean recording a local value of funds reserved  on the deal struct -- starts at 0 when deal is created, becomes ClientBalanceRequirement here, goes back to 0 after ValidateDealPublished, and then in the FailDeal handler, if reserved > 0, release those funds